### PR TITLE
Make sure that Postgres always get a consistent snapshot

### PR DESCRIPTION
**Pull request by judahrand**
## Problem

It is currently possible for PipelineWise to select an LSN to stream from which will result in missing data once the initial snapshot has been conducted.

This is due to the current mechanism of just selecting either the most recent LSN on the primary or the most recent LSN which has been replayed on the replica: https://github.com/thread/pipelinewise/blob/cfd2c2091eb60a5b71ea38486594ccfb91a4fe1c/pipelinewise/fastsync/commons/tap_postgres.py#L297-L322

This is not a safe thing to do as it does not guarantee a consistent view of the database.

## Proposed changes

Inspired by how Debezium (another CDC tool written in Java) deals with initial snapshots of Postgres databases we will take advantage of two behaviours:

1. When a replica slot is created Postgres calculates the first LSN at which the slot will provide a consistent view of the database (it also sets `confirmed_flush_lsn` to this value). We will stream from that point forward but use the current state of the database as the snapshot.
2. Since PipelineWise only receives logically decoded messages of complete transactions it should be safe to assume that any LSN that PipelineWise receives (and confirms) provides a consistent view of the database. Therefore, if the replication slot already exists we will take the current value of `confirmed_flush_lsn` to be the point at which we will start streaming.

In both of these cases, we may have data which will also be streamed to us by the replication slot but this should not be an issue as we are relying on an 'at least once' delivery mechanism already.

## Types of changes

What types of changes does your code introduce to PipelineWise?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

- [ ] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [ ] Description above provides context of the change
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Unit tests for changes (not needed for documentation changes)
- [ ] CI checks pass with my changes
- [ ] Bumping version in `setup.py` is an individual PR and not mixed with feature or bugfix PRs
- [ ] Commit message/PR title starts with `[AP-NNNN]` (if applicable. AP-NNNN = JIRA ID)
- [ ] Branch name starts with `AP-NNN` (if applicable. AP-NNN = JIRA ID)
- [ ] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [ ] Relevant documentation is updated including usage instructions

### CODE REVIEW

File: pipelinewise/fastsync/commons/tap_postgres.py [hunk 1]
```diff
@@ -1,12 +1,15 @@
+from dataclasses import dataclass
 import datetime
 import decimal
 import logging
 import re
 import sys
+from time import time
```
Comment (PeterJCLaw) [inline pipelinewise/fastsync/commons/tap_postgres.py:7]: nit: not sure if it's Pipelinewise's style to import whole modules rather than members alone, though from the other imports it looks like it might be. Certainly it's our internal style to do that, so it's probably preferable to stick to that where it's consistent with the existing code?

Comment (judahrand) [reply pipelinewise/fastsync/commons/tap_postgres.py:7]: Yup, good spot. This was just an auto VSCode import I missed!

**Comment by PeterJCLaw** [comment]
Meta: would be great if a lot of what's currently only in the PR description was also in the commit message, especially if the plan is to upstream these changes (which I'd hope we could do).

File: pipelinewise/fastsync/commons/tap_postgres.py [hunk 2]
```diff
@@ -78,11 +95,12 @@ def __get_slot_name(
 
         try:
             # Backward compatibility: try to locate existing v15 slot first. PPW <= 0.15.0
-            with connection.cursor(cursor_factory=psycopg2.extras.DictCursor) as cur:
-                cur.execute(
-                    f"SELECT * FROM pg_replication_slots WHERE slot_name = '{slot_name_v15}';"
-                )
-                v15_slots_count = cur.rowcount
+            with connection:
```
Comment (PeterJCLaw) [inline pipelinewise/fastsync/commons/tap_postgres.py:98]: Is this additional context block change intentional? What is it achieving?

File: pipelinewise/fastsync/commons/tap_postgres.py [hunk 3]
```diff
@@ -115,18 +135,23 @@ def drop_slot(cls, connection_config: Dict) -> None:
...
         finally:
             connection.close()
 
     @classmethod
-    def get_connection(cls, connection_config: Dict, prioritize_primary: bool = False):
+    def get_connection(
+        cls,
+        connection_config: Dict,
+        prioritize_primary: bool = False,
+        replication: bool = False,
```
Comment (PeterJCLaw) [inline pipelinewise/fastsync/commons/tap_postgres.py:153]: Please can we add docs on this new argument to the docstring.

**Comment by judahrand** [comment]
(I'm just re-reading through some of this... I must've been tired last night, so many typos in docstrings etc!)

File: pipelinewise/fastsync/commons/tap_postgres.py [hunk 4]
```diff
@@ -263,65 +344,51 @@ def fetch_current_log_pos(self):
...
-            else:
-                raise Exception(
-                    'Logical replication not supported before PostgreSQL 9.4'
-                )
-
-        current_lsn = result[0].get('current_lsn')
-        file, index = current_lsn.split('/')
-        lsn = (int(file, 16) << 32) + int(index, 16)
+            while start_lsn > self.get_current_lsn():
+                time.sleep(1)
```
Comment (PeterJCLaw) [inline pipelinewise/fastsync/commons/tap_postgres.py:389]: I don't think this will work? The import in this file is `from time import time`. Likely that should change to just `import time`.

Comment (PeterJCLaw) [inline pipelinewise/fastsync/commons/tap_postgres.py:389]: Is `time.sleep` is a good idea? Should there be a timeout on this loop?

File: pipelinewise/fastsync/commons/tap_postgres.py [hunk 5]
```diff
@@ -168,10 +193,15 @@ def get_connection(cls, connection_config: Dict, prioritize_primary: bool = Fals
...
+        kwargs = {}
+        if replication:
+            kwargs = {'connection_factory': psycopg2.extras.LogicalReplicationConnection}
 
-        # Set connection to autocommit
-        conn.autocommit = True
+        conn = psycopg2.connect(conn_string, **kwargs)
+
+        if not replication:
+            # Cannot set autocommit on repliication connections
```
Comment (PeterJCLaw) [inline pipelinewise/fastsync/commons/tap_postgres.py:203]:
```suggestion
            # Cannot set autocommit on replication connections
```

File: pipelinewise/fastsync/commons/tap_postgres.py [hunk 6]
```diff
@@ -115,18 +135,23 @@ def drop_slot(cls, connection_config: Dict) -> None:
...
-                    f'SELECT pg_drop_replication_slot(slot_name) '
-                    f"FROM pg_replication_slots WHERE slot_name = '{slot_name}';"
-                )
-                LOGGER.info('Number of dropped slots: %s', cur.rowcount)
-
+            with connection as conn:
+                with conn.cursor() as cur:
+                    try:
+                        cur.drop_replication_slot(slot_name)
+                        # No idea why Pylint thinks this isn't a member...
```
Comment (PeterJCLaw) [inline pipelinewise/fastsync/commons/tap_postgres.py:142]: Probably because the exceptions aren't present in the Python source (there's a warning there that "NOTE: the exceptions are injected into this module by the C extention"). Suggest updating this comment to acknowledge that?

**Comment by PeterJCLaw** [comment]
Meta: there's quite a few layered changes here which might have been clearer as separate commits.

File: pipelinewise/fastsync/commons/tap_postgres.py [hunk 7]
```diff
@@ -480,23 +547,24 @@ def copy_table(
...
-        """.format(
-            ','.join(column_safe_sql_values), schema_name, table_name
+        sql = (
+            f"COPY (SELECT {','.join(column_safe_sql_values)} "
+            ",now() AT TIME ZONE 'UTC' "
+            ",now() AT TIME ZONE 'UTC' "
+            ',null '
+            f"FROM {schema_name}.\"{table_name}\") TO STDOUT with CSV DELIMITER ','"
         )
+
```
Comment (PeterJCLaw) [inline pipelinewise/fastsync/commons/tap_postgres.py:567]: Changes seem unrelated? Is anything actually changing here?

File: pipelinewise/fastsync/commons/tap_postgres.py [hunk 8]
```diff
@@ -237,23 +266,75 @@ def create_replication_slot(self):
...
+            self.connection_config['dbname'],
+            self.connection_config['tap_id'],
+        )
+
+        res = self.primary_host_query(
+            'SELECT * '
+            'FROM pg_replication_slots '
+            f"WHERE slot_name = '{slot_name}' "
+            "AND plugin = 'wal2json' "
+            "AND slot_type = 'logical'"
```
Comment (PeterJCLaw) [inline pipelinewise/fastsync/commons/tap_postgres.py:330]: Maybe consider `f''' ... '''`?

**Comment by judahrand** [comment]
> Meta: there's quite a few layered changes here which might have been clearer as separate commits.

I can split some of it out and we can do a 'go around'?

**Comment by PeterJCLaw** [comment]
> > Meta: there's quite a few layered changes here which might have been clearer as separate commits.
> 
> I can split some of it out and we can do a 'go around'?

Yeah, that might help. Specifically I'm thinking:
- hoist `parse_lsn`
- various string layout fixes (or even just drop these as they seem unrelated)
- the changes to the context management of the connection (though again these seem unrelated?)
- extracting `get_current_lsn`
- moving to use `psycopg2`'s `create_replication_slot`
- whatever else

Not sure if that order makes sense, but broadly: split the refactors into their own smallest-atomic-commit and ensure that they're separate from the actual functional changes.

**Comment by PeterJCLaw** [comment]
Separately, I'm not sure I'm really understanding the issue, nor why this fixes it. Are there some docs I should be looking at to understand the relationship between the snaphot and the streaming? Alternatively, could you describe how they relate to each other?
I'm assuming the issue is a race condition (or equivalent) somewhere, but I don't feel I understand the system well enough to see the parts which are racing.

**Review by PeterJCLaw — COMMENTED** [review]
various comments, though if you're going to rework this they may become moot

**Comment by judahrand** [comment]
> > > Meta: there's quite a few layered changes here which might have been clearer as separate commits.
> > 
> > 
> > I can split some of it out and we can do a 'go around'?
> 
> Yeah, that might help. Specifically I'm thinking:
> 
> * hoist `parse_lsn`
> * various string layout fixes (or even just drop these as they seem unrelated)
> * the changes to the context management of the connection (though again these seem unrelated?)
> * extracting `get_current_lsn`
> * moving to use `psycopg2`'s `create_replication_slot`
> * whatever else
> 
> Not sure if that order makes sense, but broadly: split the refactors into their own smallest-atomic-commit and ensure that they're separate from the actual functional changes.

Yup, will do this. The reason it's so all over the place is that I hit a bunch of things that broken and ended up having to iterate a lot to get to this point. But absolutely see that the atomic chunks are useful.

File: pipelinewise/fastsync/commons/tap_postgres.py (hunk 8 above)
Comment (judahrand) [reply pipelinewise/fastsync/commons/tap_postgres.py:330]: The reason I did it this way is so that it doesn't show up in the logs with a bunch of weird indents and newlines etc. But I suspect there's a better middle ground 🤔

File: pipelinewise/fastsync/commons/tap_postgres.py (hunk 2 above)
Comment (judahrand) [reply pipelinewise/fastsync/commons/tap_postgres.py:98]: As it turns out I think it is now not achieving anything (though is the documented way of do using the connection): https://www.psycopg.org/docs/connection.html#the-connection-class

Can revert (or at least separate it out).

File: pipelinewise/fastsync/commons/tap_postgres.py (hunk 4 above)
Comment (judahrand) [reply pipelinewise/fastsync/commons/tap_postgres.py:389]: Silly VSCode... 😦

File: pipelinewise/fastsync/commons/tap_postgres.py (hunk 7 above)
Comment (judahrand) [reply pipelinewise/fastsync/commons/tap_postgres.py:567]: Not really. Just again trying to get the SQL formatted in a way which doesn't make the logs look horrific. But can not include here.

File: pipelinewise/fastsync/commons/tap_postgres.py (hunk 6 above)
Comment (judahrand) [reply pipelinewise/fastsync/commons/tap_postgres.py:142]: Ah! That makes sense.

File: pipelinewise/fastsync/commons/tap_postgres.py (hunk 4 above)
Comment (judahrand) [reply pipelinewise/fastsync/commons/tap_postgres.py:389]: Yup, there probably should.
